mlflow/types/schema.py

```diff
@@ -122,34 +122,270 @@ def from_numpy_type(cls, np_type):
+        optional `required` keys.
+        """
+        if not {"name", "type"} <= set(kwargs.keys()):
+            raise MlflowException(
+                "Missing keys in Property JSON. Expected to find keys `name` and `type`"
+            )
+        name = kwargs.pop("name")
+        required = kwargs.pop("required", True)
+        dtype = kwargs["type"]
+        if dtype not in ["array", "object"]:
+            return cls(name=name, dtype=dtype, required=required)
+        if dtype == "array":
+            return cls(name=name, dtype=Array.from_json_dict(**kwargs), required=required)
+        if dtype == "object":
+            return cls(name=name, dtype=Object.from_json_dict(**kwargs), required=required)
+
+
+@experimental
+class Object:
+    """
+    Specification used to represent a json-convertible object.
+    """
+
+    def __init__(self, properties: List[Property]) -> None:
+        self._check_properties(properties)
+        self._properties = properties
+
+    def _check_properties(self, properties):
+        if not isinstance(properties, list):
+            raise MlflowException.invalid_parameter_value(
+                f"Expected properties to be a dict, got type {type(properties).__name__}"
+            )
+        if any(not isinstance(v, Property) for v in properties):
+            raise MlflowException.invalid_parameter_value(
+                "Expected values to be instance of Property"
+            )
+        # check duplicated property names
+        names = set()
+        duplicates = set()
+        for prop in properties:
+            if prop.name in names:
+                duplicates.add(prop.name)
+            else:
+                names.add(prop.name)
+        if len(duplicates) > 0:
+            raise MlflowException.invalid_parameter_value(
+                f"Found duplicated property names: {duplicates}"
+            )
+
+    @property
+    def properties(self) -> List[Property]:
+        """The list of object properties"""
+        return self._properties
+
+    @properties.setter
+    def properties(self, value: List[Property]) -> None:
+        self._check_properties(value)
+        self._properties = value
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, Object):
+            return sorted(self.properties, key=lambda x: x.name) == sorted(
+                other.properties, key=lambda x: x.name
+            )
+        return False
+
+    def to_dict(self):
+        properties = sorted([prop.to_dict() for prop in self.properties], key=lambda x: x["name"])
+        return {
+            "type": "object",
+            # Sort by name to make sure the order is stable
+            "properties": properties,
+            # Question: should we update as below (which aligns with the design doc)
+            # but it makes to_dict & from_json_dict not symmetric
+            # "properties": {prop.pop("name"): prop for prop in properties},
+        }
```

Comment:
The difference here is:
"properties": [{"name": "content", "type": "string", "required": True}]
versus
"properties": {"content": {"type": "string", "required": True}}

Reply:
I prefer the symmetrical approach. Are there any tradeoffs to other portions of this larger project to making that change?

Reply:
Nope, I think it only affects to_dict and from_json_dict. The second representation needs Property's to_dict to extract name out as a key. Actually after rethinking it's also symmetric, but ColSpec.to_dict() also uses `d["name"] = self.name`, am wondering if we want to be consistent here.

Reply:
I think consistency will make the code easier to maintain and reason about :)

Reply:
Synced with Corey, we'll follow JSONSchema's pattern https://json-schema.org/learn/getting-started-step-by-step#nest-data to extract keys out :)